src/remoted/secure.c

```diff
@@ -386,22 +416,28 @@ static void HandleSecureMessage(char *buffer, int recv_b, struct sockaddr_in *pe
             key_unlock();
             mwarn(DENYIP_WARN, srcip);
 
+            // Send key request by ip
+            push_request(srcip,"ip");
             if (sock_client >= 0)
                 _close_sock(&keys, sock_client);
 
             return;
+        } else {
+            ip_found = 1;
         }
+
         tmp_msg = buffer;
     }
 
     /* Decrypt the message */
-
-    tmp_msg = ReadSecMSG(&keys, tmp_msg, cleartext_msg, agentid, recv_b - 1, &msg_length, srcip);
-
-    if (tmp_msg == NULL) {
-
+    if (ReadSecMSG(&keys, tmp_msg, cleartext_msg, agentid, recv_b - 1, &msg_length, srcip, &tmp_msg) != KS_VALID) {
         /* If duplicated, a warning was already generated */
         key_unlock();
+        if (ip_found) {
```

Comment:
If the agent sent an ID number (_ip_found=0_), Remoted should also request a key, but in this case it should find an agent by ID.